src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -35311,6 +35330,92 @@ static void Main()
         }
 
+        [Fact]
+        public void UnannotatedConstraint_Override()
+        {
+            var source0 =
+@"using System.Runtime.CompilerServices;
+public interface I<T> { }
+public abstract class A<T> where T : class
+{
+    [NonNullTypes(false)] public abstract void F1<U>() where U : T, I<T>;
+    [NonNullTypes(false)] public abstract void F2<U>() where U : T?, I<T?>;
+    [NonNullTypes(true)] public abstract void F3<U>() where U : T, I<T>;
+    [NonNullTypes(true)] public abstract void F4<U>() where U : T?, I<T?>;
+}";
+            var comp0 = CreateCompilation(new[] { source0, NonNullTypesAttributesDefinition }, parseOptions: TestOptions.Regular8);
+            comp0.VerifyDiagnostics();
+            var ref0 = comp0.EmitToImageReference();
+
+            var source =
+@"using System.Runtime.CompilerServices;
+[NonNullTypes(false)]
+class B1 : A<string>
+{
+    public override void F1<U>() { }
+    public override void F2<U>() { }
+    public override void F3<U>() { }
+    public override void F4<U>() { }
+}
+[NonNullTypes(false)]
+class B2 : A<string?>
+{
+    public override void F1<U>() { }
+    public override void F2<U>() { }
+    public override void F3<U>() { }
+    public override void F4<U>() { }
+}
+[NonNullTypes(true)]
+class B3 : A<string>
+{
+    public override void F1<U>() { }
+    public override void F2<U>() { }
+    public override void F3<U>() { }
+    public override void F4<U>() { }
+}
+[NonNullTypes(true)]
+class B4 : A<string?>
+{
+    public override void F1<U>() { }
+    public override void F2<U>() { }
+    public override void F3<U>() { }
+    public override void F4<U>() { }
+}";
+            var comp = CreateCompilation(source, references: new[] { new CSharpCompilationReference(comp0) }, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+            verifyAllConstraintTypes();
+
+            comp = CreateCompilation(source, references: new[] { comp0.EmitToImageReference() }, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+            verifyAllConstraintTypes();
+
+            void verifyAllConstraintTypes()
+            {
+                verifyConstraintTypes("B1.F1", "System.String", "I<System.String>");
+                verifyConstraintTypes("B1.F2", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B1.F3", "System.String", "I<System.String>");
+                verifyConstraintTypes("B1.F4", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F1", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F2", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F3", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F4", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B3.F1", "System.String!", "I<System.String!>");
```

Comment:
From our discussion, it's not obvious what should happen when you substitute a `string!` into `T~`.

Reply:
Added question to spec-let.

---
In reply to: [199652696](https://github.com/dotnet/roslyn/pull/28248#discussion_r199652696) [](ancestors = 199652696)